File: trunk/src/m/plot/processmesh.m
function [x y z elements is2d isplanet]=processmesh(md,data,options)
%PROCESSMESH - process mesh to be plotted
%
%   Usage:
%      [x y z elements is2d]=processmesh(md,data,options)
%
%   See also: PLOTMODEL, PROCESSDATA

%some checks
if md.mesh.numberofvertices==0,
	error('plot error message: mesh is empty')
end
if md.mesh.numberofvertices==md.mesh.numberofelements
	error(['plot error message: the number of elements is the same as the number of nodes...']);
end

%special case for mesg 2dvertical
if strcmp(domaintype(md.mesh),'2Dvertical'),
	[x y z elements is2d isplanet] = processmesh(md.mesh,options);
	return;
end

%special case for mesh 3dsurface
if strcmp(domaintype(md.mesh),'3Dsurface'),
	[x y z elements is2d isplanet] = processmesh(md.mesh,options);
	return;
end

if ~strcmpi(getfieldvalue(options,'coord','xy'),'latlon'),
	x=md.mesh.x;
	if isprop(md.mesh,'x2d'), x2d=md.mesh.x2d; end
	y=md.mesh.y;
	if isprop(md.mesh,'y2d'), y2d=md.mesh.y2d; end
else
	x=md.mesh.long;
	y=md.mesh.lat;
end

if isprop(md.mesh,'z'),
	z=md.mesh.z;
else
	z=zeros(size(x));
end
z=getfieldvalue(options,'z',z);
if ischar(z),
	z=md.(z);
end

if isprop(md.mesh,'elements2d'), elements2d=md.mesh.elements2d; end
elements=md.mesh.elements;

%is it a 2d plot?
if md.mesh.dimension()==2,
	is2d=1;
else
	if getfieldvalue(options,'layer',0)>=1,
		is2d=1;
	else
		is2d=0;
	end
end

%layer projection? 
if getfieldvalue(options,'layer',0)>=1,
	if strcmpi(getfieldvalue(options,'coord','xy'),'latlon'),
		error('processmesh error message: cannot work with 3D meshes for now');
	end
	%we modify the mesh temporarily to a 2d mesh from which the 3d mesh was extruded. 
	x=x2d;
	y=y2d;
	z=zeros(size(x2d));
	elements=elements2d;
end

%units
if exist(options,'unit'),
	unit=getfieldvalue(options,'unit');
	x=x*unit;
	y=y*unit;
	z=z*unit;
end

if isa(md,'planet'),
	isplanet=1;
else
	isplanet=0;
end

File: trunk/src/py3/plot/processmesh.py
from math import isnan
import MatlabFuncs as m
import numpy as np

def processmesh(md,data,options):
	"""
	PROCESSMESH - process the mesh for plotting

	Usage:
		x,y,z,elements,is2d=processmech(md,data,options)

	See also: PLOTMODEL, PROCESSDATA
	"""

	#some checks
	if md.mesh.numberofvertices==0:
		raise ValueError('processmesh error: mesh is empty')
	if md.mesh.numberofvertices==md.mesh.numberofelements:
		raise ValueError('processmesh error: the number of elements is the same as the number of nodes')

	if len(data)==0 or not isinstance(data,dict):
		
		if 'latlon' not in options.getfieldvalue('coord','xy').lower(): #convert to lower case for comparison
			x=md.mesh.x
			if 'x2d' in dir(md.mesh): x2d=md.mesh.x2d
			y=md.mesh.y
			if 'y2d' in dir(md.mesh): y2d=md.mesh.x2d
		else:
			x=md.mesh.long
			y=md.mesh.lat

		if 'z' in dir(md.mesh):
			z=md.mesh.z
		else:
			z=np.zeros_like(md.mesh.x)
		
		if 'elements2d' in dir(md.mesh): 
			elements2d=md.mesh.elements2d
			elements2d=elements2d-1  # subtract one since python indexes from zero
		elements=md.mesh.elements
		elements=elements-1

		#is it a 2D plot?
		if md.mesh.dimension()==2:
			is2d=1
		else:
			if options.getfieldvalue('layer',0)>=1:
				is2d=1
			else:
				is2d=0

		#layer projection?
		if options.getfieldvalue('layer',0)>=1:
			 if 'latlon' in options.getfieldvalue('coord','xy').lower():
				 raise ValueError('processmesh error: cannot work with 3D mesh in lat-lon coords')
			#we modify the mesh temporarily to a 2D mesh from which the 3D mesh was extruded
			 x=x2d
			 y=y2d
			 z=zeros(size(x2d))
			 elements=elements2d
	
	else:
		#Process mesh for plotting 
		if md.mesh.dimension()==2:
			is2d=1
		else:
			# process polycollection here for 3D plot
			is2d=0
	
	#units
	if options.exist('unit'):
		unit=options.getfieldvalue('unit')
		x=x*unit
		y=y*unit
		z=z*unit

	#is model a member of planet class? (workaround until planet class defined)
	if md.__class__.__name__!='model':
		isplanet=1
	else:
		isplanet=0

	return x,y,z,elements,is2d,isplanet
